### ISSUE TITLE
You should not include Windows specific headers when building under Cygwin

### ISSUE THREAD

**Issue by nanis** [Cyan4973/xxHash#100]
In [xxhsum.c][1], you define some Windows specific constructs when building under Cygwin. That is not correct. The purpose of Cygwin is to provide a POSIX API layer over Windows APIs (whereas the purpose of MinGW* projects is to allow Windows programs to be compiled using the GCC toolchain).

```text
$ gcc -DXXH_PRIVATE_API -O3 -march=native xxhsum.c -o xxhsum-cygwin.exe
xxhsum.c: In function ‘BMK_hash’:
xxhsum.c:71:42: warning: implicit declaration of function ‘_fileno’ [-Wimplicit-function-declaration]
 #  define SET_BINARY_MODE(file) _setmode(_fileno(file), _O_BINARY)
                                          ^
xxhsum.c:523:9: note: in expansion of macro ‘SET_BINARY_MODE’
         SET_BINARY_MODE(stdin);
         ^
xxhsum.c:71:57: error: ‘_O_BINARY’ undeclared (first use in this function)
 #  define SET_BINARY_MODE(file) _setmode(_fileno(file), _O_BINARY)
                                                         ^
xxhsum.c:523:9: note: in expansion of macro ‘SET_BINARY_MODE’
         SET_BINARY_MODE(stdin);
         ^
xxhsum.c:71:57: note: each undeclared identifier is reported only once for each function it appears in
 #  define SET_BINARY_MODE(file) _setmode(_fileno(file), _O_BINARY)
                                                         ^
xxhsum.c:523:9: note: in expansion of macro ‘SET_BINARY_MODE’
         SET_BINARY_MODE(stdin);
         ^
xxhsum.c: In function ‘main’:
xxhsum.c:72:33: warning: implicit declaration of function ‘_isatty’ [-Wimplicit-function-declaration]
 #  define IS_CONSOLE(stdStream) _isatty(_fileno(stdStream))
                                 ^
xxhsum.c:1226:33: note: in expansion of macro ‘IS_CONSOLE’
     if ( (filenamesStart==0) && IS_CONSOLE(stdin) ) return badusage(exename);
```

With the following change:

```diff
$ git diff
diff --git a/xxhsum.c b/xxhsum.c
index 1928141..322cfd1 100644
--- a/xxhsum.c
+++ b/xxhsum.c
@@ -62,7 +62,7 @@
 /*-************************************
 *  OS-Specific Includes
 **************************************/
-#if defined(MSDOS) || defined(OS2) || defined(WIN32) || defined(_WIN32) || defined(__CYGWIN__)
+#if defined(MSDOS) || defined(OS2) || defined(WIN32) || defined(_WIN32)
 #  include <fcntl.h>    /* _O_BINARY */
 #  include <io.h>       /* _setmode, _isatty */
 #  ifdef __MINGW32__
```

We get:

```text
$ gcc -DXXH_PRIVATE_API -O3 -march=native xxhsum.c -o xxhsum-cygwin
$ ./xxhsum-cygwin -h
./xxhsum-cygwin 0.6.2 (64-bits little endian), by Yann Collet
...
```

HTH.

[1]: https://github.com/Cyan4973/xxHash/blob/7fc5ce8592f10c68506f4a070bfa6a9ab2004b44/xxhsum.c#L65

**Comment by t-mat** [comment]
Hi @nanis, thanks for the heads up!
@Cyan4973, using [LZ4's programs/platform.h](https://github.com/lz4/lz4/blob/dev/programs/platform.h) would be better for this issue.

**Comment by t-mat** [comment]
I don't try it in Cygwin.
But @nanis, could you try to compile LZ4 in your Cygwin environment, and provide a version number of Cygwin which you are using?
I suppose LZ4's platform.h also may cause error in your environment.

----

Cygwin had provided mingw compatible [`_isatty()`](http://www.cygwin.com/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=winsup/mingw/include/io.h;h=259e5088a0b9c339191da30d80a0235511a40d86;hb=1aca2cb5f4a7cc79169ec4fe4b53f94cd6e01994#l369) and [`_setmode()`](http://www.cygwin.com/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=winsup/mingw/include/io.h;h=259e5088a0b9c339191da30d80a0235511a40d86;hb=1aca2cb5f4a7cc79169ec4fe4b53f94cd6e01994#l347)
But they [got rid of it](http://www.cygwin.com/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=winsup/cygwin/include/io.h;h=9525f6ba053a32a0ab205df4a9cff9b95d493536;hb=c9e4b69e9f48a517ec5c6bbf20c80bd744117b14#l22) for some reason.  They also completely removed winsup/mingw/.

**Comment by Cyan4973** [comment]
This line, with `__CYGWIN__` detection macro,
was added 2 years ago in a bid to add `stdin` input method to `xxhsum`.

It was likely copy/pasted from somewhere else, probably a combination of `lz4io.c` and `lz4cli.c`.
Whatever it was, it's no longer visible, as it has been replaced by `platform.h`.

`platform.h` continues using `__CYGWIN__` macro detection for the purpose of defining `IS_CONSOLE()` : 
https://github.com/lz4/lz4/blob/dev/programs/platform.h#L111

and it's still combined with `MSDOS`, `_WIN32` and such.

The solution proposed (removed `__CYGWIN__` macro detection) looks simple enough.
But it requires proper testing to ensure we don't trigger problems that made this macro detection there in the first place.

I currently have no cygwin configuration around, and unsure how to build one.

**Comment by cup** [comment]
https://github.com/Cyan4973/xxHash/issues/100
https://github.com/nim-lang/Nim/issues/10645

**Comment by Cyan4973** [comment]
I made a test compiling `xxhsum` on a recent `cygwin` installation.
Using the exact command line listed in this issue : 
```
gcc -DXXH_PRIVATE_API -O3 -march=native xxhsum.c -o xxhsum-cygwin.exe`
```
compilation works perfectly, with no warning nor error,
while `__CYGWIN__` code path is still present in source code.

I went further and analyzed how compilation works on `cygwin`.
And actually, the `__CYGWIN__` code path does no longer matter : `POSIX` path comes first, and is selected during compilation of `xxhsum` under `cygwin`.

Which also means that this `__CYGWIN__` code path (coupled with `DOS` and `OS2`) is not useful. So it can be removed.
